### PR TITLE
Adding Instructional Resources to Reporting UI 

### PR DESCRIPTION
**Pull request by dsnyder-fairway**
### Overall view of a group
screen shot 2017-12-08 at 1 11 13 pm

### Showing links (how are we supposed to grab the URL link text?)
screen shot 2017-12-08 at 5 04 43 pm

### Student score distribution (no links)
screen shot 2017-12-08 at 1 11 28 pm

### Datatable (no links)
screen shot 2017-12-08 at 1 11 35 pm

###  Changes to student history
screen shot 2017-12-08 at 1 11 47 pm
screen shot 2017-12-08 at 1 11 54 pm

### CODE REVIEW

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html [hunk 1]
```diff
@@ -48,26 +47,24 @@
...
                     </span>
-              <span class="icon"><i class="fa" [ngClass]="{'fa-minus':session.filter, 'fa-plus': !session.filter }"></i></span>
             </a>
           </div>
         </div>
       </div>
 
       <!-- Aggregate Scale Score -->
-      <div class="col-md-7">
+      <div class="col-md-9 right">
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html]: what's right? `text-right` `pull-right`?

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts [hunk 2]
```diff
@@ -213,6 +221,21 @@ export class AssessmentResultsComponent implements OnInit {
...
+    this.instructionalResourcesService.getInstructionalResources(this.assessmentExam.assessment.id, this.assessmentProvider.getSchoolId()).subscribe((instructionalResources: InstructionalResources) => {
+      let resources = instructionalResources.getResourcesByPerformance(performanceLevel);
+      if (resources.length === 0) {
+        this.content = this.translateService.instant('labels.groups.results.assessment.no-instruct-found');
+      }
+
+      resources.forEach(resource => {
+        this.content = this.content.concat('<p>' + resource.url + '</p>');
+      });
+      // this.content = this.content.concat('<p><a [href]="' + this.sanitizer.bypassSecurityTrustUrl(resource.url) + '/>"</p>');
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: `<a [attr.href]="url">` doesn't do it?

File: webapp/src/main/webapp/src/app/assessments/results/view/results-by-student/results-by-student.component.ts [hunk 3]
```diff
@@ -93,13 +96,29 @@ export class ResultsByStudentComponent implements OnInit {
...
+  loadInstructionalResources(exam: Exam) {
+    this.content = '';
+    this.instructionalResourcesService.getInstructionalResources(this.assessment.id, exam.school.id).subscribe((instructionalResources: InstructionalResources) => {
+      let resources = instructionalResources.getResourcesByPerformance(exam.level);
+      if (resources.length === 0) {
+        this.content = this.translate.instant('labels.groups.results.assessment.no-instruct-found');
+      }
+
+      resources.forEach(resource => {
+        this.content = this.content.concat('<p>' + resource.url + '</p>');
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/view/results-by-student/results-by-student.component.ts]: i think this appears about 3 times - we missing some reuse opportunity?

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/view/results-by-student/results-by-student.component.ts]: yes -- I had trouble figuring out how to extract this method correctly

File: webapp/src/main/webapp/src/app/shared/button/button.component.html [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        outsideClick="true">
+  <div *ngIf="icon">
+    <span class="sr-only">{{'labels.more-info' | translate}}</span>
+    <i class="fa {{ icon }}"></i>
+  </div>
+</button>
+
+<ng-template #buttonPopover>
+  <div [innerHTML]="content"></div>
+</ng-template>
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/shared/button/button.component.html]: should we call this `popover-button` maybe? seems like more than just a button

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 2 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: chrome devtools displays
`WARNING: sanitizing HTML stripped some content (see http://g.co/ng/security#xss).` after setting that

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: can you debug log the value before it is written? sounds like the value actually had something invalid

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts [hunk 5]
```diff
@@ -92,6 +95,9 @@ export class AssessmentResultsComponent implements OnInit {
   @Input()
   minimumItemDataYear: number;
 
+  @Input()
+  content: string;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: this is pretty confusing naming. what is this the content for? usually we don't pass around html content like this.
think you can use <ng-template #templateName> and then pass a template ref or something.

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 2 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: screen shot 2017-12-08 at 2 04 34 pm

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 5 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: I have some similar concerns about the [innerHTML] above.  I'd much rather provide data than HTML.

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 2 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: I think this may be related to using [innerHTML]  Have you tried putting the HTML into the .html and just providing an array of models?

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html [hunk 6]
```diff
@@ -1,77 +1,111 @@
 <!-- IAB display -->
-<div *ngIf="showIab && hasAverageScore" class="row">
+<div *ngIf="(showIab || showIcaSummative) && hasAverageScore" class="row">
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: I think (showIab || showIcaSummative) is always *true* isn't it?

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html [hunk 7]
```diff
@@ -1,77 +1,111 @@
...
+            </div>
+          </td>
+        </tr>
+      </table>
     </div>
   </div>
 </div>
+<ng-template #buttonPopover>
+  <div [innerHTML]="content"></div>
+</ng-template>
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: Since the button template is duplicated here, I'd be tempted to create an InstructionalResourcePopoverComponent and just do:
```
<ng-template #buttonPopover>
  <instructional-resource-popover [resources]="instructionalResources" />
</ng-template>
```

File: webapp/src/main/webapp/src/app/assessments/results/instructional-resources.mapper.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+const organizationLevels: string[] = [
+  "System",
+  "State",
+  "DistrictGroup",
+  "District",
+  "SchoolGroup"
+];
+
+@Injectable()
+export class InstructionalResourcesMapper {
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/instructional-resources.mapper.ts]: Since this mapper is only used by one service, I'd be tempted to move the mapping methods into the service itself and get rid of this injectable mapper class

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html (hunk 6 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: it should be. This was done to get rid of some duplication with setting the `average-scale-score-display`. I'll remove the tautology

File: webapp/src/main/webapp/src/app/shared/button/information-button.component.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * This component is responsible for displaying a label with
+ * an information popover icon.
+ */
+@Component({
+  selector: 'info-button,[info-button]',
+  templateUrl: './information-button.component.html'
+})
+export class InformationButtonComponent extends ButtonComponent {
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/shared/button/information-button.component.ts]: This breakout confuses me a little bit, it doesn't seem to be much of a code-saver?

File: webapp/src/main/webapp/src/app/shared/button/button.component.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * This component is responsible for displaying a button with
+ * a specified popover icon.
+ */
+@Component({
+  selector: 'sb-button,[sb-button]',
+  templateUrl: './button.component.html'
+})
+export class ButtonComponent {
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/shared/button/button.component.ts]: It doesn't look like this is being used anywhere?

File: webapp/src/main/webapp/src/app/shared/button/information-button.component.ts (hunk 9 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/shared/button/information-button.component.ts]: Yeah, to some extent this is an artifact of a previous PR I made for this where I had a BookButtonComponent extending the ButtonComponent as well. I was unsure if I should have undone all of my changes to these files (was previous `InformationLinkComponent`)

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html (hunk 6 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: done

File: webapp/src/main/webapp/src/app/shared/button/information-button.component.ts (hunk 9 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/shared/button/information-button.component.ts]: done

File: webapp/src/main/webapp/src/app/shared/button/button.component.ts (hunk 10 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/shared/button/button.component.ts]: removed

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html (hunk 1 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html]: corrected (`pull-right`)

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html]: i think with columns its not advised to use `float` (`pull-right`) you probably want to put a div in the column with `pull-right`

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.html]: I think it was that there was another `row` appearing within a `row`. I removed the `pull-right` and the additional `row` in c646013

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html (hunk 7 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: done

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 5 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: removed in 9b98a76. Links are displaying correctly now

File: webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts (hunk 2 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/assessment-results.component.ts]: this has been resolved in 9b98a76

**Comment by dsnyder-fairway** [comment]
URLs are now displaying correctly 
screen shot 2017-12-08 at 5 04 43 pm

**Comment by dsnyder-fairway** [comment]
@wtritch cceed62 adds changes for display names

File: webapp/src/main/webapp/src/assets/i18n/en.json [hunk 11]
```diff
@@ -377,6 +377,18 @@
...
+      "link": {
+        "System": "Smarter Balanced Instructional Resource",
+        "State": " Instructional Resource",
+        "DistrictGroup": "{{organizationName}} Instructional Resource",
+        "District": "{{organizationName}} Instructional Resource",
+        "SchoolGroup": "{{organizationName}} Instructional Resource"
+      }
+    },
+    "states": {
+      "CA": "California"
```
Comment (esotrope) [inline webapp/src/main/webapp/src/assets/i18n/en.json]: @wtritch we already have this info in admin config repo settings right?

File: webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html [hunk 12]
```diff
@@ -1,7 +1,9 @@
 <div *ngIf="resources !== undefined && resources.length > 0; else noResources">
   <p class="popover-link"
      *ngFor="let instructionalResource of resources">
-    <a [attr.href]="instructionalResource.url" target="_blank">{{ instructionalResource.url }}</a>
+    <a *ngIf="instructionalResource.organizationLevel === 'System'" [attr.href]="instructionalResource.url" target="_blank">{{ 'labels.instructional-resources.link.System' | translate }}</a>
+    <a *ngIf="instructionalResource.organizationLevel === 'State'" [attr.href]="instructionalResource.url" target="_blank">{{ ('labels.states.' + instructionalResource.stateCode) | translate }}{{ 'labels.instructional-resources.link.State' | translate }}</a>
+    <a *ngIf="instructionalResource.organizationLevel !== 'System' && instructionalResource.organizationLevel !== 'State'" [attr.href]="instructionalResource.url" target="_blank">{{ ('labels.instructional-resources.link.' + instructionalResource.organizationName) | translate:instructionalResource }}</a>
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html]: i think i'd be tempted to have one line for the anchor and pre-process this stuff before sending it to the view

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html]: 👍 agreed. I extracted this into a `linkText` method

File: webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html [hunk 13]
```diff
@@ -1,9 +1,7 @@
 <div *ngIf="resources !== undefined && resources.length > 0; else noResources">
   <p class="popover-link"
-     *ngFor="let instructionalResource of resources">
-    <a *ngIf="instructionalResource.organizationLevel === 'System'" [attr.href]="instructionalResource.url" target="_blank">{{ 'labels.instructional-resources.link.System' | translate }}</a>
-    <a *ngIf="instructionalResource.organizationLevel === 'State'" [attr.href]="instructionalResource.url" target="_blank">{{ ('labels.states.' + instructionalResource.stateCode) | translate }}{{ 'labels.instructional-resources.link.State' | translate }}</a>
-    <a *ngIf="instructionalResource.organizationLevel !== 'System' && instructionalResource.organizationLevel !== 'State'" [attr.href]="instructionalResource.url" target="_blank">{{ ('labels.instructional-resources.link.' + instructionalResource.organizationName) | translate:instructionalResource }}</a>
+     *ngFor="let instructionalResource of resources; let i = index">
+    <a [attr.href]="instructionalResource.url" target="_blank">{{ linkText(i) }}</a>
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html]: does this have to be `[attr.href]` anymore? i think it can be `<a href="{{instructionalResource.url}}"`

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html]: also - why not pass the `instructionalResource` variable so you can do away with the lookup by index?

File: webapp/src/main/webapp/src/assets/i18n/en.json (hunk 11 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/assets/i18n/en.json]: @wtritch heads up - I decided on using `stateCode` here instead of `stateName` here as we could desire translation for state names (for instance, NY - Nueva York)

File: webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html (hunk 13 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/popover/instructional-resource-popover.component.html]: done

**Comment by jtreuting** [comment]
looking at the UI screenshots, I see the links listed but not as described in the mockups on the requirements page.  It should be saying something like "California Resources", "District A Resources", etc.

**Comment by dsnyder-fairway** [comment]
@jtreuting sorry those changes were included in PRs since cceed62. Here is an example
screen shot 2017-12-12 at 9 37 54 am

**Comment by wtritch** [comment]
hmm... you should never have two "California" resources displayed in one popup

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/instructionalresource/InstructionalResource.java [hunk 14]
```diff
@@ -31,6 +32,13 @@ public String getOrganizationName() {
         return organizationName;
     }
 
+    /**
+     * @return The state code this resource is for.
+     */
+    public String getStateCode() {
+        return stateCode;
+    }
```
Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/instructionalresource/InstructionalResource.java]: I wouldn't add this.  I'd saturate the organizationName property with the "app.state.name" configuration value IFF the organizationLevel == State. (At the service layer.)

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/instructionalresource/JdbcInstructionalResourceRepository.java [hunk 15]
```diff
@@ -18,6 +18,9 @@
     @Value("${sql.instructionalResource.findAllByAssessmentId}")
     private String findAllByAssessmentId;
 
+    @Value("${app.state.code}")
+    private String stateCode;
```
Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/instructionalresource/JdbcInstructionalResourceRepository.java]: I recommend injecting the ApplicationProperties instance rather than an @Value injection

**Comment by dsnyder-fairway** [comment]
@wtritch - I believe this is due to some SQL I ran to generate this example:
`insert into reporting.instructional_resource (asmt_name, resource, performance_level, org_level) (select name,  'https://portal.smarterbalanced.org/library/en/digital-library-connections-high-school-brief-writes.docx', 0, 'System' from reporting.asmt);` and then modified performance level. Probably not realistic

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html [hunk 16]
```diff
@@ -1,77 +1,111 @@
...
+            <span class="pull-right">
+              {{ examLevelEnum + performanceLevel.id.toString() | translate }}
+              <button
+                class="btn btn-borderless btn-xs icon-only pull-right"
+                angulartics2On="click"
+                angularticsEvent="InstructionalResource"
+                angularticsCategory="AssessmentResults"
+                [disabled]="false"
+                [popover]="buttonPopover"
+                popoverTitle="{{ 'labels.groups.results.assessment.instruct-button-prefix' | translate }}{{ examLevelEnum + performanceLevel.id.toString() | translate }}"
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.html]: I don't recommend breaking a single statement into two translations, it assumes the sentance/statement structure.  Recommend consolidating into a single translation (a little duplication is ok, rather than assuming the value will always be divided into left and right segments)

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts [hunk 17]
```diff
@@ -2,6 +2,12 @@ import { Component, Input } from "@angular/core";
 import { AssessmentExam } from "../model/assessment-exam.model";
 import { ExamStatistics, ExamStatisticsLevel } from "../model/exam-statistics.model";
 import { ScaleScoreService } from "./scale-score.service";
+import { InstructionalResource, InstructionalResources } from "../model/instructional-resources.model";
+import { InstructionalResourcesService } from "./instructional-resources.service";
+import { GroupAssessmentService } from "../../groups/results/group-assessment.service";
+
+const icaColors = [ 'maroon', 'gray-darkest', 'green-dark', 'blue-dark' ];
+const iabColors = [ 'blue-dark', 'blue-dark aqua', 'aqua' ];
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: Why is this here?  Don't we define these colors elsewhere?  @esotrope ?  Aren't we using an indexed color provider service?

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts [hunk 18]
```diff
@@ -63,7 +74,39 @@ export class AverageScaleScoreComponent {
       return this.statistics.levels;
   }
 
-  getLevelPercent(num: number): number {
-    return this.levelPercents[num];
+  levelSum(): number {
+    const values = this.statistics.levels.map(l => l.value);
+    return values.reduce((p, c) => p + c);
+  }
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: I feel like there are a lot of orthogonal changes creeping into this PR.  I don't recommend trying to break everything out now, but in the future try to keep a single PR to a single task.  (Sometimes even multiple PRs per task in order to keep things focused/small/readable. =)

File: webapp/src/main/webapp/src/app/assessments/results/instructional-resources.mapper.ts (hunk 8 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/assessments/results/instructional-resources.mapper.ts]: ...Still recommend removing this mapper =)

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts (hunk 17 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: These were found [here](https://github.com/SmarterApp/RDW_Reporting/pull/646/files/fa67212a8bb66aeb05ee9664c047b56bb1f28ac8#diff-4061477d62dc8d21f8086960477b2b99). Someone suggested I refactor this with a for-loop

File: webapp/src/main/webapp/src/assets/i18n/en.json (hunk 11 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/assets/i18n/en.json]: We'll only support a single state per one installation.  I'd argue that it's easier to provide the state code and name together as configuration properties rather than as translation overrides in production when this system is deployed by multiple states.  I'd also argue that the state name should probably not be translated into spanish, vietnamese, etc as a pronoun.

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts (hunk 17 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: i think we have a ColorService among the webapp angular components - yes. we can see if that satisfies the desired results. What are these colors for btw?

File: webapp/src/main/webapp/src/assets/i18n/en.json (hunk 11 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/assets/i18n/en.json]: @esotrope yup, although the property naming should be consistent to allow setting it via a shared application.yml in the configuration repo: "app.state.code and app.state.name"

File: webapp/src/main/webapp/src/assets/i18n/en.json [hunk 19]
```diff
@@ -366,10 +367,27 @@
...
       "students": "Students",
-      "student-score-distribution": "Student Score Distribution"
+      "student-score-distribution": {
+        "title": "Student Score Distribution"
+      }
+    },
+    "instructional-resources": {
+      "link": {
+        "System": "Smarter Balanced Instructional Resource",
+        "State": " Instructional Resource",
```
Comment (wtritch) [inline webapp/src/main/webapp/src/assets/i18n/en.json]: This is what I was referring to above, assuming that a statement will have the same structure in different languages has caused me problems on projects in the past.  I'd recommend changing this to:
`"{{organizationName}} Instructional Resource"`

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts (hunk 18 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: also, just in general let's stay away from single letter variable names - they don't communicate what is what here. The only exception to this is `i` in a for loop i think

File: webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts (hunk 17 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: These colors are for providing different score distribution colors

Comment (wtritch) [reply webapp/src/main/webapp/src/app/assessments/results/average-scale-score.component.ts]: My only real goal here is that we don't want these colors defined in multiple ways in multiple locations.  Because colors are something everybody has an opinion on and I could see them changing =P

**Comment by dsnyder-fairway** [comment]
I'll be re-opening this once I've wrapped all of this up
